docs/developer/decision-records/2025-04-25-agreement-policy-scope/README.md

```diff
@@ -0,0 +1,77 @@
+# Contract Agreement policy scope
+
+## Decision
+
+We will add a `contract.agreement` policy scope, that will be evaluated by the provider before the automatic agreement.
+
+## Rationale
+
+The negotiation flow is at the moment an automatic one: a negotiation can happen if the `contract.negotiation` policy 
+scope evaluation succeeds, or it gets terminated if the evaluation fails.
+
+To implement "manual negotiation approval", the `PendingGuard` service can be used, but this doesn't permit to
+
+## Approach
+
+There will be a new policy scope/context to permit the new `PolicyEngine` evaluation, the class will look like:
+
+```java
+public class ContractAgreementPolicyContext extends PolicyContextImpl {
+
+    @PolicyScope
+    public static final String AGREEMENT_SCOPE = "contract.agreement";
+
+    public ContractAgreementPolicyContext() {
+        this.agent = agent;
+    }
+
+    @Override
+    public String scope() {
+        return AGREEMENT_SCOPE;
+    }
+}
+```
+
+An additional evaluation of the `PolicyEngine` will be added in the `ProviderContractNegotiationManager` in the
+`processRequested` method, so that every negotiation that gets `REQUESTED` can be pass through policy functions:
+```java
+    @WithSpan
+    private boolean processRequested(ContractNegotiation negotiation) {
+        var evaluation = policyEngine.evaluate(negotiation.getLastContractOffer().getPolicy(), new ContractAgreementPolicyContext());
+        if (evaluation.succeeded()) {
+            transitionToAgreeing(negotiation);
+        } else {
+            negotiation.setPending(true);
+            observable.invokeForEach(l -> l.manualInteractionRequired());
+            update(negotiation);
+        }
+    }
+```
+
+By doing this in the case the evaulation goes through the behavior will be the same (automatic agreement), otherwise the
+negotiation will be put as "pending", that's the flag that prevent it to be taken into consideration from the state
+machine.
+The `observable` call will be used to generate an event, so that the client could get notified about the negotiation
+waiting for manual interaction.
+
+The policy function could be bind to an ODRL `Duty` constraint in the `Permission` with the [`reviewPolicy` `Action`](https://www.w3.org/TR/odrl-vocab/#term-reviewPolicy).
+Example:
+```java
+{
+  "@context": "http://www.w3.org/ns/odrl.jsonld",
+  "@type": "Offer",
+  "permission": [{
+    "action": "use",
+    "duty": [{
+      "action": "reviewPolicy"
+    }]
+  }]
+}
```

Comment:
I like the idea of having an out-of-the-box available policy function for this functionality, but when putting this as a duty into the policy, wouldn't that technically (in ODRL terms) mean that the consumer has to review the policy (since when initiating a negotiation, the provider is set as the `assigner` and the consumer is set as the `assignee`)?

Reply:
Uhm... I'm definitely no expert on this, but your comment makes total sense.
what could be an alternative?

Reply:
maybe basing the decision whether to set the negotiation to `pending` on the presence of specific attributes on the asset or policy? ID (or private properties directly) of asset or policy could be passed in the policy context for this scope. The provider would then need a way to configure for which properties the negotiation should be set to pending, e.g. via a configuration or an API to have it more dynamic. The policy function would check whether any of the configured properties is present on the policy/asset and return false if it is. The provider could then simply flag assets/policies accordingly.

Reply:
tbh the odrl model is too vague and looks lke the same result can be achieved with different rule actions, so I just removed this part, no default policy function will be there, it will be up to adopters to implement their own